File: ios/CallPilot/Protocol/ContentModels.swift
import Foundation

enum MessageDirection: String, Codable, CaseIterable {
    case inbound = "INBOUND"
    case outbound = "OUTBOUND"
}

enum MessageDeliveryStatus: String, Codable, CaseIterable {
    case received = "RECEIVED"
    case sent = "SENT"
    case failed = "FAILED"
    case error = "ERROR"
}

struct SMSMessage: Codable, Equatable, Hashable, Identifiable {
    let messageId: String
    let revision: String
    let direction: MessageDirection
    let address: String
    let text: String
    let occurredAt: Int64
    let recordedAt: Int64
    let status: MessageDeliveryStatus

    var id: String { messageId }

    private static var messageIdRE: Regex<Substring> { /^msg_[A-Za-z0-9_-]{12,80}$/ }
    private static var revisionRE: Regex<Substring> { /^revision_[A-Za-z0-9_-]{12,80}$/ }

    init(
        messageId: String,
        revision: String,
        direction: MessageDirection,
        address: String,
        text: String,
        occurredAt: Int64,
        recordedAt: Int64,
        status: MessageDeliveryStatus
    ) {
        self.messageId = messageId
        self.revision = revision
        self.direction = direction
        self.address = address
        self.text = text
        self.occurredAt = occurredAt
        self.recordedAt = recordedAt
        self.status = status
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let messageId = try values.decode(String.self, forKey: .messageId)
        let revision = try values.decode(String.self, forKey: .revision)
        let direction = try values.decode(MessageDirection.self, forKey: .direction)
        let address = try values.decode(String.self, forKey: .address)
        let text = try values.decode(String.self, forKey: .text)
        let occurredAt = try values.decode(Int64.self, forKey: .occurredAt)
        let recordedAt = try values.decode(Int64.self, forKey: .recordedAt)
        let status = try values.decode(MessageDeliveryStatus.self, forKey: .status)

        guard messageId.wholeMatch(of: Self.messageIdRE) != nil,
              revision.wholeMatch(of: Self.revisionRE) != nil,
              occurredAt >= 0,
              recordedAt >= 0,
              status != .received || direction == .inbound else {
            throw DecodingError.dataCorruptedError(
                forKey: .messageId,
                in: values,
                debugDescription: "Invalid content-sync message"
            )
        }

        self.init(
            messageId: messageId,
            revision: revision,
            direction: direction,
            address: address,
            text: text,
            occurredAt: occurredAt,
            recordedAt: recordedAt,
            status: status
        )
    }
}

struct MessagePage: Codable, Equatable {
    let v: Int
    let items: [SMSMessage]
    let nextCursor: String?
    let hasMore: Bool
    let collectionRevision: String
    let oldestAvailableAt: Int64?

    private static var revisionRE: Regex<Substring> { /^revision_[A-Za-z0-9_-]{12,80}$/ }

    init(
        v: Int,
        items: [SMSMessage],
        nextCursor: String?,
        hasMore: Bool,
        collectionRevision: String,
        oldestAvailableAt: Int64?
    ) {
        self.v = v
        self.items = items
        self.nextCursor = nextCursor
        self.hasMore = hasMore
        self.collectionRevision = collectionRevision
        self.oldestAvailableAt = oldestAvailableAt
    }

    init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let v = try values.decode(Int.self, forKey: .v)
        let items = try values.decode([SMSMessage].self, forKey: .items)
        let nextCursor = try values.decodeIfPresent(String.self, forKey: .nextCursor)
        let hasMore = try values.decode(Bool.self, forKey: .hasMore)
        let collectionRevision = try values.decode(String.self, forKey: .collectionRevision)
        let oldestAvailableAt = try values.decodeIfPresent(Int64.self, forKey: .oldestAvailableAt)

        guard v == 1,
              items.count <= 100,
              Set(items.map(\.messageId)).count == items.count,
              hasMore == (nextCursor != nil),
              ContentWireValidation.validCursor(nextCursor),
              collectionRevision.wholeMatch(of: Self.revisionRE) != nil,
              oldestAvailableAt == nil || oldestAvailableAt! >= 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .v,
                in: values,
                debugDescription: "Invalid content-sync message page"
            )
        }

        self.init(
            v: v,
            items: items,
            nextCursor: nextCursor,
            hasMore: hasMore,
            collectionRevision: collectionRevision,
            oldestAvailableAt: oldestAvailableAt
        )
    }
}

@MainActor
protocol MessageContentClient: AnyObject {
    func listMessages(limit: Int, cursor: String?) async throws -> MessagePage
}

File: android/app/src/main/kotlin/ai/bondings/callpilot/protocol/ContentModels.kt
package ai.bondings.callpilot.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json

@Serializable
enum class MessageDirection { INBOUND, OUTBOUND }

@Serializable
enum class MessageDeliveryStatus { RECEIVED, SENT, FAILED, ERROR }

@Serializable
data class SMSMessage(
    val messageId: String,
    val revision: String,
    val direction: MessageDirection,
    val address: String,
    val text: String,
    val occurredAt: Long,
    val recordedAt: Long,
    val status: MessageDeliveryStatus,
) {
    fun requireValid() {
        checkContract(ContentWireValidation.validMessageId(messageId))
        checkContract(ContentWireValidation.validRevision(revision))
        checkContract(occurredAt >= 0 && recordedAt >= 0)
        checkContract(status != MessageDeliveryStatus.RECEIVED || direction == MessageDirection.INBOUND)
    }
}

@Serializable
data class MessagePage(
    val v: Int,
    val items: List<SMSMessage>,
    val nextCursor: String?,
    val hasMore: Boolean,
    val collectionRevision: String,
    val oldestAvailableAt: Long?,
) {
    fun requireValid() {
        checkContract(v == 1)
        checkContract(items.size <= 100)
        items.forEach(SMSMessage::requireValid)
        checkContract(items.map(SMSMessage::messageId).toSet().size == items.size)
        checkContract(hasMore == (nextCursor != null))
        checkContract(ContentWireValidation.validCursor(nextCursor))
        checkContract(ContentWireValidation.validRevision(collectionRevision))
        checkContract(oldestAvailableAt == null || oldestAvailableAt >= 0)
    }

    companion object {
        fun decode(json: Json, text: String): MessagePage = try {
            json.decodeFromString<MessagePage>(text).also(MessagePage::requireValid)
        } catch (error: ContentContractException) {
            throw error
        } catch (error: SerializationException) {
            throw ContentContractException(error)
        } catch (error: IllegalArgumentException) {
            throw ContentContractException(error)
        }
    }
}

interface MessageContentClient {
    suspend fun listMessages(limit: Int, cursor: String?): MessagePage
}

object ContentWireValidation {
    private val CURSOR = Regex("^cursor_[A-Za-z0-9_-]+$")
    private val MESSAGE_ID = Regex("^msg_[A-Za-z0-9_-]{12,80}$")
    private val CALL_ID = Regex("^call_[A-Za-z0-9_-]{12,80}$")
    private val TIMELINE_ITEM_ID = Regex("^item_[A-Za-z0-9_-]{12,80}$")
    private val REVISION = Regex("^revision_[A-Za-z0-9_-]{12,80}$")
    private val PRODUCT_CODE = Regex("^[A-Z][A-Z0-9_]{2,63}$")

    fun validCursor(value: String?): Boolean =
        value == null || (value.length <= 2_048 && CURSOR.matches(value))

    fun validRevision(value: String): Boolean = REVISION.matches(value)

    fun validMessageId(value: String): Boolean = MESSAGE_ID.matches(value)

    fun validCallId(value: String): Boolean = CALL_ID.matches(value)

    fun validTimelineItemId(value: String): Boolean = TIMELINE_ITEM_ID.matches(value)

    fun validProductCode(value: String): Boolean = PRODUCT_CODE.matches(value)
}

class ContentContractException(cause: Throwable? = null) : Exception("Invalid content-sync payload", cause)

private fun checkContract(condition: Boolean) {
    if (!condition) throw ContentContractException()
}
